File: web/models/conversation-collaboration.ts
import type { AgentBrief, Conversation } from '@/models/conversation'

export type CollaborationOnlineStatus = 'online' | 'busy' | 'offline'
export type CollaborationInvitationStatus = 'pending' | 'accepted' | 'declined' | 'canceled' | 'expired'

export type CollaborationTarget = {
  id: number
  name: string
  display_name: string | null
  job_number: string | null
  avatar: string | null
  online_status: CollaborationOnlineStatus
  current_count: number
  max_concurrent: number
  available: boolean
  disabled_reason: 'already_joined' | 'pending' | 'offline' | null
}

export type CollaborationTargetListResponse = {
  items: CollaborationTarget[]
  total: number
}

export type CollaborationInvitation = {
  id: number
  conversation_id: number
  status: CollaborationInvitationStatus
  inviter: AgentBrief | null
  invitee: AgentBrief | null
  owner: AgentBrief | null
  visitor_name: string | null
  channel_name: string | null
  last_message_preview: string | null
  expires_at: string
  responded_at: string | null
  created_at: string
}

export type CollaborationInvitationListResponse = {
  items: CollaborationInvitation[]
  total: number
}

export type CollaborationInvitationRespondResponse = {
  invitation: CollaborationInvitation
  conversation: Conversation | null
}

File: server/app/schemas/conversation_collaboration.py
"""
Conversation collaboration schemas.
"""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.schemas.conversation import AgentBrief, ConversationResponse


class CollaborationTarget(BaseModel):
    id: int
    name: str
    display_name: str | None = None
    job_number: str | None = None
    avatar: str | None = None
    online_status: Literal["online", "busy", "offline"]
    current_count: int
    max_concurrent: int
    available: bool
    disabled_reason: str | None = None


class CollaborationTargetListResponse(BaseModel):
    items: list[CollaborationTarget]
    total: int


class CollaborationInvitationCreate(BaseModel):
    invitee_id: int


class CollaborationInvitationRespond(BaseModel):
    action: Literal["accept", "decline"]


class CollaborationInvitationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    conversation_id: int
    status: str
    inviter: AgentBrief | None = None
    invitee: AgentBrief | None = None
    owner: AgentBrief | None = None
    visitor_name: str | None = None
    channel_name: str | None = None
    last_message_preview: str | None = None
    expires_at: datetime
    responded_at: datetime | None = None
    created_at: datetime


class CollaborationInvitationListResponse(BaseModel):
    items: list[CollaborationInvitationResponse]
    total: int


class CollaborationInvitationRespondResponse(BaseModel):
    invitation: CollaborationInvitationResponse
    conversation: ConversationResponse | None = None
